File: 257.binary-tree-paths/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        if (root == null) return Collections.emptyList();
        List<String> list = new ArrayList<String>();
        binaryTreePaths("", root, list);
        return list;
    }
    
    public void binaryTreePaths(String s, TreeNode root, List<String> list) {
        if (!s.isEmpty()) {
            s += "->";
        }
        s += root.val;
        if (root.left == null && root.right == null) list.add(s);
        if (root.left != null) binaryTreePaths(s, root.left, list);
        if (root.right != null) binaryTreePaths(s, root.right, list);
    }
}

File: 257.binary-tree-paths/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if root is None:
            return []
        if root.left is None and root.right is None:
            return [str(root.val)]
        result = []
        if root.left is not None:
            result0 = self.binaryTreePaths(root.left)
            for s in result0:
                result.append(str(root.val) + '->' + s)
        if root.right is not None:
            result0 = self.binaryTreePaths(root.right)
            for s in result0:
                result.append(str(root.val) + '->' + s)
        return result
